### ISSUE TITLE
Access multiple return values in multicast delegates

### ISSUE THREAD

**Issue by sirinath** [dotnet/runtime#49119]
One drawback with the current implementation of delegates is that multiple return values cannot be accessed. Can the language/api be enhanced so that multiple return values can be accessed.

**Comment by huoyaoyuan** [comment]
The better design would be remove multicast at all and use a list of delegates.
Once considering exceptions, it will be as complex as using list or array.

**Comment by stephentoub** [comment]
> One drawback with the current implementation of delegates is that multiple return values cannot be accessed. Can the language/api be enhanced so that multiple return values can be accessed.

Just to make sure it's clear, you can, you just need to get the invocation list and invoke the delegates manually:
```C#
using System;

Func<int> func = null;

func += () => 1;
func += () => 2;
func += () => 3;

foreach (Func<int> f in func.GetInvocationList())
{
    Console.WriteLine(f());
}
```
prints
```
1
2
3
```

**Comment by sirinath** [comment]
What I was hoping was that delegates would be rethought and redesigned in perhaps a backward compatible way without way avoiding their current drawbacks.

**Comment by stephentoub** [comment]
What are the specific drawbacks you're concerned about?  What is your proposal for addressing them?  At the moment this isn't actionable.

**Comment by sirinath** [comment]
Perhaps and Invoke overload which returns a list of either the return value or any thrown exception would solve the issue. Also syntactic sugar to use this would be a  bonus.

**Comment by huoyaoyuan** [comment]
I don't think this is a good idea. Multicast by default is already a mistake, and we shouldn't continue and encourage that.

**Comment by stephentoub** [comment]
FWIW, you can do this via your own extension method, e.g.
```C#
public static TResult[] InvokeAll<TResult>(this Func<TResult> func)
{
    Delegate[] delegates = func.GetInvocationList();
    var results = new TResult[delegates.Length];
    var exceptions = new List<Exception>();
    for (int i = 0; i < delegates.Length; i++)
    {
        try
        {
            results[i] = ((Func<TResult>)delegates[i])();
        }
        catch (Exception e)
        {
            exceptions.Add(e);
        }
    }

    if (exceptions.Count > 0)
    {
        throw new AggregateException(exceptions);
    }

    return results;
}
```
